### ISSUE TITLE
Control key not being held down in headless mode

### ISSUE THREAD

**Issue by olliechick** [TestFX/TestFX#566]
I am using TestFX to test selecting multiple items in a ListView using the Ctrl key, then clicking a button, which generates an alert dialog whose content is based on the items selected.

```
press(KeyCode.CONTROL);

clickOn((Node) lookup(hasText("Item 0")).query());
verifyThat((Node) lookup(hasText("Item 0")).query(), Node::isFocused);

clickOn((Node) lookup(hasText("Item 1")).query());
verifyThat((Node) lookup(hasText("Item 1")).query(), Node::isFocused);

clickOn((Node) lookup(hasText("Item 2")).query());
verifyThat((Node) lookup(hasText("Item 2")).query(), Node::isFocused);

release(KeyCode.CONTROL);

clickOn("actionButton");
alertDialogHasHeaderAndContent("Items selected: 3"); // my own function
```
### In headful mode

The tests all pass.

### In headless mode

The three `verifyThat` tests all pass, but when it clicks on the button and verifies that the resulting alert dialog is correct, it isn't; the alert generated is as if only one item from the list was selected (i.e., expected behaviour if the control key was not held down).

It results in:
````
org.junit.ComparisonFailure: 
Expected :Items selected: 3
Actual   :Items selected: 1
````
----

Here is `alertDialogHasHeaderAndContent` for completeness:
```
/**
 * Checks the current alert dialog displayed (on the top of the window stack) has the expected contents.
 *
 * From https://stackoverflow.com/a/48654878/8355496
 * Licenced under cc by-sa 3.0 with attribution required https://creativecommons.org/licenses/by-sa/3.0/
 * @param expectedHeader Expected header of the dialog
 * @param expectedContent Expected content of the dialog
 */
private void alertDialogHasHeaderAndContent(final String expectedHeader, final String expectedContent) {
    final Stage actualAlertDialog = getTopModalStage();
    assertNotNull(actualAlertDialog);

    final DialogPane dialogPane = (DialogPane) actualAlertDialog.getScene().getRoot();
    assertEquals(expectedHeader, dialogPane.getHeaderText());
    assertEquals(expectedContent, dialogPane.getContentText());
}

/**
 * Get the top modal window.
 *
 * Adapted from https://stackoverflow.com/a/48654878/8355496
 * Licenced under cc by-sa 3.0 with attribution required https://creativecommons.org/licenses/by-sa/3.0/
 * @return the top modal window
 */
private Stage getTopModalStage() {
    // Get a list of windows but ordered from top[0] to bottom[n] ones.
    // It is needed to get the first found modal window.
    final List<Window> allWindows = new ArrayList<>(new FxRobot().robotContext().getWindowFinder().listWindows());
    Collections.reverse(allWindows);

    return (Stage) allWindows
            .stream()
            .filter(window -> window instanceof Stage)
            .findFirst()
            .orElse(null);
}
```

**Comment by brcolow** [comment]
Thanks for the report. I do have a suggestion if you are interested in getting this fixed. The first thing I would recommend is to create a failing test case in OpenJFX upstream. This feels like an upstream bug to me because key state is handled by the Glass robot implementation in JavaFX, but I could be wrong. If you can create such a failing test, there is a good chance the issue will be looked into by the OpenJFX devs, or at least they can help point us in the right direction.

If you fork and clone the https://github.com/javafxports/openjdk-jfx repository you will see the directory:

`tests/system/src/test/java/test/robot/com/sun/glass/ui/monocle`

Inside that directory are some tests such as `DoubleClickTest`, `DragAndDropTest`, etc. The goal would be to create the simplest possible test-case that demonstrates that CTRL is not being held down through multiple clicks.

I personally have never written such a test so I am not sure if you will run into any show-stoppers but taking a quick browse of the existing tests shows that there is quite a plethora to base new tests on (copy and paste ftw).

Sorry that this involves work for you but it's the best I can currently do to help you.

**Comment by brcolow** [comment]
Please keep me posted if you make any progress. If it isn't reproducible on upstream then we can take a closer look at TestFX.

**Comment by MatthewSmit** [comment]
I believe I've found the issue, com.sun.glass.ui.monocle.MouseState doesn't appear to use the MODIFIER_CONTROL (and other keyboard modifier) values when sending the events to javafx.

**Comment by brcolow** [comment]
Do you mean these lines:

https://github.com/javafxports/openjdk-jfx/blob/96bbea07f2675af93212af1a097628814f86680e/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MouseInput.java#L159

https://github.com/javafxports/openjdk-jfx/blob/96bbea07f2675af93212af1a097628814f86680e/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MouseInput.java#L108

https://github.com/javafxports/openjdk-jfx/blob/96bbea07f2675af93212af1a097628814f86680e/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MouseInput.java#L148

https://github.com/javafxports/openjdk-jfx/blob/96bbea07f2675af93212af1a097628814f86680e/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MouseInput.java#L179

https://github.com/javafxports/openjdk-jfx/blob/96bbea07f2675af93212af1a097628814f86680e/modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/MouseInput.java#L197

I think you are right.

**Comment by MatthewSmit** [comment]
Yes, those lines. I've created a patch to use locally, which solves our immediate issue.
I'm struggling with creating a bug report with openjfx, since it appears you need to be given an account first. I've created a bug report here, [https://bugreport.java.com/bugreport/](https://bugreport.java.com/bugreport/) and hopefully oracle will forward it to the appropriate location.

**Comment by brcolow** [comment]
I would encourage you to open a PR on the https://github.com/javafxports/openjdk-jfx repository with a link to this issue and your bug report number.
